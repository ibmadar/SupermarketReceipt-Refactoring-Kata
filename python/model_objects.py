from enum import Enum


class Product:
    def __init__(self, name, unit):
        self.name = name
        self.unit = unit

    def __str__(self):
        return f"{self.name} (units:{self.unit.name.lower()})"


class ProductUnit(Enum):
    EACH = 1
    KILO = 2


class SpecialOfferType(Enum):
    THREE_FOR_TWO = 1
    TEN_PERCENT_DISCOUNT = 2
    TWO_FOR_AMOUNT = 3
    FIVE_FOR_AMOUNT = 4

class Offer:
    def __init__(self, offer_type, product, argument=None):
        self.offer_type = offer_type
        self.product = product

    def __str__(self):
        return f"\n{self.offer_type.name.lower()} {self.product.name}"
    
    def calculate_discount(self, unit_price, quantity):
        raise NotImplementedError("Subclasses should implement this!")
        
class Ten_percent_offer(Offer):
    def __init__(self, product, discount):
        super().__init__(SpecialOfferType.TEN_PERCENT_DISCOUNT, product)
        self.discount = discount

    def __str__(self):
        return f"\n{self.offer_type.name.lower()} {self.product.name} {self.discount}%"

    def calculate_discount(self, unit_price, quantity):
        return Discount(self.product, str(self.discount) + "% off", -quantity * unit_price * self.discount / 100.0)    


class Units_for_amount(Offer):
    def __init__(self, offertype, product, units, amount):
        super().__init__(offertype, product)
        self.amount = amount
        self.units = units

    def calculate_discount(self, unit_price, quantity):
        discount_n = unit_price * quantity - (quantity // self.units) * self.amount - (quantity % self.units) * unit_price
        
        return Discount(self.product, f"Buy {self.units} for ${self.amount}", -discount_n)

    def __str__(self):
        return f"\n{self.offer_type.name.lower()} {self.product.name} ${self.argument}"

class Five_for_amount_offer(Units_for_amount):
    def __init__(self, product, amount):
        super().__init__(SpecialOfferType.FIVE_FOR_AMOUNT, product, 5, amount)

class Two_for_amount_offer(Units_for_amount):
    def __init__(self, product, amount):
        super().__init__(SpecialOfferType.TWO_FOR_AMOUNT, product, 2, amount)
        
    
class Three_for_two_offer(Offer):
    def __init__(self, product):
        super().__init__(SpecialOfferType.THREE_FOR_TWO, product)

    def calculate_discount(self, unit_price, quantity):
        discount_n = unit_price * quantity - (quantity // 3) * unit_price * 2

        return Discount(self.product, f"Buy 3 for ${unit_price * 2}", -discount_n)
    
    def __str__(self):
        return f"\n{self.offer_type.name.lower()} {self.product.name}"    




class Discount:
    def __init__(self, product, description, discount_amount):
        self.product = product
        self.description = description
        self.discount_amount = discount_amount

    def __str__(self):
        return f" Discount: Prod:{self.product} [Desc:{self.description}  Disc. Amount:{str(self.discount_amount)}]"


class OfferFactory:
    @staticmethod
    def create_offer(offer_type, product, argument):
        match offer_type:
            case SpecialOfferType.THREE_FOR_TWO:
                return Three_for_two_offer(product)
            case SpecialOfferType.TEN_PERCENT_DISCOUNT:
                return Ten_percent_offer(product, argument)
            case SpecialOfferType.TWO_FOR_AMOUNT:
                return Two_for_amount_offer(product, argument)
            case SpecialOfferType.FIVE_FOR_AMOUNT:
                return Five_for_amount_offer(product, argument) 
            case _:
                raise ValueError("Unknown offer type")