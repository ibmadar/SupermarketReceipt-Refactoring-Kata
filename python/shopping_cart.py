import math

from model_objects import SpecialOfferType, Discount


class ShoppingCart:

    def __init__(self):
        self._items = {}

    @property
    def items(self):
        return self._items.items()

    def add_item(self, product, quantity=1):
        self._items[product] = self._items.get(product,0) + quantity

    



                
