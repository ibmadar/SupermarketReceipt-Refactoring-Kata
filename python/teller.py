from model_objects import Offer, OfferFactory
from receipt import Receipt


class Teller:

    def __init__(self, catalog):
        self.catalog = catalog
        self.offers = {}

    
    def add_special_offer(self, offer_type, product, argument=None):
        self.offers[product] = OfferFactory.create_offer(offer_type, product, argument)

    def checks_out_articles_from(self, the_cart):
        receipt = Receipt()
        for prod, quantity in the_cart.items:
            unit_price = self.catalog.unit_price(prod)
            receipt.add_product(prod, quantity, unit_price)
            offer = self.offers.get(prod,None)
            if offer:
                receipt.add_discount(offer.calculate_discount(unit_price, quantity))
        return receipt

    
