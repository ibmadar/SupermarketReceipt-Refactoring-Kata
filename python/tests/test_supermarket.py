import pytest
import approvaltests

from model_objects import Product, SpecialOfferType, ProductUnit
from shopping_cart import ShoppingCart
from teller import Teller
from tests.fake_catalog import FakeCatalog


def test_regular_shopping():
    catalog = FakeCatalog()

    #Products
    toothbrush = Product("toothbrush", ProductUnit.EACH)
    apples = Product("apples", ProductUnit.KILO)

    #Add them to catalog
    catalog.add_product(toothbrush, 0.99)    
    catalog.add_product(apples, 1.99)

    teller = Teller(catalog)

   
    #Shopping cart
    cart = ShoppingCart()
    cart.add_item(apples, 2.5)
    cart.add_item(toothbrush, 2)

    receipt = teller.checks_out_articles_from(cart)

    approvaltests.verify(receipt)

def test_ten_percent_discount():
    catalog = FakeCatalog()

    #Products
    toothbrush = Product("toothbrush", ProductUnit.EACH)
    apples = Product("apples", ProductUnit.KILO)

    #Add them to catalog
    catalog.add_product(toothbrush, 0.99)    
    catalog.add_product(apples, 1.99)

    teller = Teller(catalog)

    #Add offers
    teller.add_special_offer(SpecialOfferType.TEN_PERCENT_DISCOUNT, toothbrush, 10.0)

    #Shopping cart
    cart = ShoppingCart()
    cart.add_item(apples, 2.5)
    cart.add_item(toothbrush, 2)

    receipt = teller.checks_out_articles_from(cart)

    approvaltests.verify(receipt)


def test_same_item__added_multiple_times():
    catalog = FakeCatalog()

    #Products
    toothbrush = Product("toothbrush", ProductUnit.EACH)
    apples = Product("apples", ProductUnit.KILO)

    #Add them to catalog
    catalog.add_product(toothbrush, 0.99)    
    catalog.add_product(apples, 1.99)

    teller = Teller(catalog)

    #Add offers
    teller.add_special_offer(SpecialOfferType.TEN_PERCENT_DISCOUNT, toothbrush, 10.0)

    #Shopping cart
    cart = ShoppingCart()
    cart.add_item(apples, 2.5)
    cart.add_item(apples, 2)

    receipt = teller.checks_out_articles_from(cart)

    approvaltests.verify(receipt)    

def test_no_offers():
    catalog = FakeCatalog()

    #Products
    toothbrush = Product("toothbrush", ProductUnit.EACH)
    apples = Product("apples", ProductUnit.KILO)

    #Add them to catalog
    catalog.add_product(toothbrush, 0.99)    
    catalog.add_product(apples, 1.99)

    teller = Teller(catalog)


    #Shopping cart
    cart = ShoppingCart()
    cart.add_item(apples, 2.5)
    cart.add_item(toothbrush, 2)

    receipt = teller.checks_out_articles_from(cart)

    approvaltests.verify(receipt)    


def test_3x2_discount():
    catalog = FakeCatalog()

    #Products
    toothbrush = Product("toothbrush", ProductUnit.EACH)
    apples = Product("apples", ProductUnit.KILO)

    #Add them to catalog
    catalog.add_product(toothbrush, 2)    
    catalog.add_product(apples, 1.99)

    teller = Teller(catalog)

    #Add offers
    teller.add_special_offer(SpecialOfferType.THREE_FOR_TWO, toothbrush)

    #Shopping cart
    cart = ShoppingCart()
    cart.add_item(apples, 2.5)
    cart.add_item(toothbrush, 3)

    receipt = teller.checks_out_articles_from(cart)

    approvaltests.verify(receipt)    


def test_2_for_amount_discount():
    catalog = FakeCatalog()

    #Products
    toothbrush = Product("toothbrush", ProductUnit.EACH)
    apples = Product("apples", ProductUnit.KILO)

    #Add them to catalog
    catalog.add_product(toothbrush, 2)    
    catalog.add_product(apples, 1.99)

    teller = Teller(catalog)

    #Add offers
    teller.add_special_offer(SpecialOfferType.TWO_FOR_AMOUNT, toothbrush,1.5)

    #Shopping cart
    cart = ShoppingCart()
    cart.add_item(apples, 2.5)
    cart.add_item(toothbrush, 3)

    receipt = teller.checks_out_articles_from(cart)

    approvaltests.verify(receipt)        