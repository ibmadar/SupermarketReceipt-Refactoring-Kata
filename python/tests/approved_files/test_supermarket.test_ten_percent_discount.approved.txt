Receipt:
Item: apples (units:kilo) - Qty:2.5 - Unit price:1.99 - Price:4.975
Item: toothbrush (units:each) - Qty:2 - Unit price:0.99 - Price:1.98 Discount: Prod:toothbrush (units:each) [Desc:10.0% off  Disc. Amount:-0.198]
Total: 6.757
